test/integration-tests/lib/minerd/process.rs

```diff
@@ -0,0 +1,670 @@
+use std::{
+    fs,
+    net::SocketAddr,
+    path::PathBuf,
+    sync::{Arc, Mutex},
+};
+use tokio::{
+    io::{AsyncBufReadExt, AsyncReadExt, AsyncWriteExt, BufReader},
+    net::{TcpListener, TcpStream},
+    process::{Child as TokioChild, Command as TokioCommand},
+};
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, info};
+
+use crate::utils::{http, tarball};
+
+use super::error::MinerdError;
+
+const VERSION_MINERD: &str = "2.5.1";
+
+fn get_minerd_filename(os: &str, arch: &str) -> Result<String, MinerdError> {
+    match (os, arch) {
+        ("macos", "aarch64") => Ok(format!(
+            "pooler-cpuminer-{VERSION_MINERD}-arm64-apple-darwin.tar.gz"
+        )),
+        // todo
+        // ("macos", "x86_64") =>
+        // "pooler-cpuminer-{VERSION_MINERD}-x86_64-apple-darwin.tar.gz",
+        ("linux", "x86_64") => Ok(format!(
+            "pooler-cpuminer-{VERSION_MINERD}-x86_64-linux-gnu.tar.gz"
+        )),
+        ("linux", "aarch64") => Ok(format!(
+            "pooler-cpuminer-{VERSION_MINERD}-aarch64-linux-gnu.tar.gz"
+        )),
+        _ => Err(MinerdError::OsArchNotSupported(format!(
+            "OS or architecture not supported: {} {}",
+            os, arch
+        ))),
+    }
+}
+
+/// A wrapper struct for the minerd process that provides:
+/// - TCP proxy functionality to intercept communications
+/// - Process management for spawning and killing minerd
+#[derive(Debug)]
+pub struct MinerdProcess {
+    /// Path to the minerd binary
+    minerd_binary: PathBuf,
+    /// Handle to the spawned minerd process
+    process: Arc<Mutex<Option<TokioChild>>>,
+    /// Address where the wrapper listens for minerd connections
+    local_address: SocketAddr,
+    /// Address of the upstream mining server
+    upstream_address: SocketAddr,
+    /// Whether to kill the process after the first mining.submit
+    single_submit: bool,
+    /// Cancellation token to coordinate shutdown of all tasks
+    cancellation_token: CancellationToken,
+}
+
+impl MinerdProcess {
+    /// Creates a new MinerdProcess with the given upstream address
+    pub async fn new(
+        upstream_address: SocketAddr,
+        single_submit: bool,
+    ) -> Result<Self, MinerdError> {
+        let current_dir: PathBuf = std::env::current_dir().expect("failed to read current dir");
+        let minerd_dir = current_dir.join("minerd");
+
+        let os = std::env::consts::OS;
+        let arch = std::env::consts::ARCH;
+        let download_filename = get_minerd_filename(os, arch)?;
+
+        if !minerd_dir.exists() {
+            fs::create_dir_all(minerd_dir.clone()).expect("failed to create minerd directory");
+            let download_endpoint = format!(
+                "https://github.com/plebhash/cpuminer/releases/download/v{VERSION_MINERD}/"
```

Comment:
ideally this should be
```suggestion
                "https://github.com/stratum-mining/cpuminer/releases/download/v{VERSION_MINERD}/"
```